### PR TITLE
Avadino

### PR DESCRIPTION
**Pull request by avavita**
...fixed a billion things.

### CODE REVIEW

File: dino_catalog/Dinodex_Wizard.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    input = STDIN.gets.split("=", 2)
+
+    abort("You aren't paying attention") unless input.count == 2
+
+    input.each do |x|
+      x.strip!
+    end
+
+  	return input if SUPPORTED_FIELDS.include? input[0]
```
Comment (avavita) [inline dino_catalog/Dinodex_Wizard.rb]: oh damn it sublime

File: dino_catalog/Dinodex_Wizard.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  def self.prompt_user_for_constraint
+    puts "Valid fields listed below:\n#{SUPPORTED_FIELDS}"
+    puts "Please type a field to search and the value to look for, separated by ="
+    puts "{ex: name = velociraptor}"
+
+    input = STDIN.gets.split("=", 2)
+
+    abort("You aren't paying attention") unless input.count == 2
+
+    input.each do |x|
```
Comment (zbright) [inline dino_catalog/Dinodex_Wizard.rb]: Good idea hound!

File: dino_catalog/dinoConverter.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  "Weight" => "WEIGHT_IN_LBS", 
+  "Walking" => "WALKING"
+  }
+
+  def self.csv_file_to_dino_array(file)
+    csv = CSV::parse(File.open(file, 'r') {|f| f.read })
+    fields = csv.shift
+    hash_array = csv.map { |record| Hash[*fields.zip(record).flatten] }
+    fixed_headers_array = hash_array.map { |hash| fix_pirate_hash(hash) }
+    dino_array = []
```
Comment (zbright) [inline dino_catalog/dinoConverter.rb]: check out `each_eith_object` for something like this

File: dino_catalog/dinoConverter.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    end
+    dino_array
+  end
+
+  def self.csv_files_to_dino_array(files)
+    abort("No files given") if files == []
+    arrays = files.map do |file|
+      csv_file_to_dino_array(file)
+    end
+    dino_array = []
```
Comment (zbright) [inline dino_catalog/dinoConverter.rb]: `each_with_object`

File: dino_catalog/dinoConverter.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    arrays.each do |array|
+      dino_array += array
+    end
+    dino_array
+  end
+
+  def self.fix_pirate_hash(hash)
+    if hash.keys.include? "Genus"
+      fixed_h = Hash[hash.map {|k, v| [MAPPINGS[k], v] }]
+      case fixed_h["DIET"]
```
Comment (zbright) [inline dino_catalog/dinoConverter.rb]: this case seems to be over kill. Maybe a ternary instead?

`fiexed_h['DIET'] == 'Yes' ? 'Carnivore' : 'Unknown'`

File: dino_catalog/dinoConverter.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    end
+    dino_array = []
+    arrays.each do |array|
+      dino_array += array
+    end
+    dino_array
+  end
+
+  def self.fix_pirate_hash(hash)
+    if hash.keys.include? "Genus"
```
Comment (zbright) [inline dino_catalog/dinoConverter.rb]: invert this if with a return to simplify. 
`return hash unless hash.keys.include? "Genus"`

File: dino_catalog/dino.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    @period = attr_hash["PERIOD"]
+    @continent = attr_hash["CONTINENT"]
+    @diet = attr_hash["DIET"]
+    @weight_in_lbs = attr_hash["WEIGHT_IN_LBS"]
+    @walking = attr_hash["WALKING"]
+    @description = attr_hash["DESCRIPTION"]
+  end
+
+  def like?(field, value)
+    if respond_to? field.downcase
```
Comment (zbright) [inline dino_catalog/dino.rb]: in general, if you find yourself returning from an if statement its best to invert it and have an explicit return

File: dino_catalog/dinodex.rb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  require_relative 'dinoConverter'
+  require 'pp'
+
+  BIG = 4000
+
+  def initialize(files)
+    @array_of_dinos = DinoConverter.csv_files_to_dino_array(files)
+  end
+
+  def array_of_dinos
```
Comment (zbright) [inline dino_catalog/dinodex.rb]: just make this variable readable - let ruby do the magic :)

`attr_reader array_of_dinos`

File: dino_catalog/dino.rb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      (send field.downcase).include? value
```
Comment (zbright) [inline dino_catalog/dino.rb]: Why use include here? I may just be tired, but from what I can tell, this is calling one of the variables defined in the initializer. In which case, there would be only one value returned? 

Also, comparing strings without (at least) downcasing scares me

File: dino_catalog/dinoConverter.rb [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+      fixed_h = Hash[hash.map {|k, v| [MAPPINGS[k], v] }]
+      case fixed_h["DIET"]
+        when "Yes"
+          fixed_h["DIET"] = "Carnivore"
+        else
+          fixed_h["DIET"] = "Unknown"
```
Comment (zbright) [inline dino_catalog/dinoConverter.rb]: Instead of unknown, go with 'Herbivore'. Its just one of of the Joe-isms

File: dino_catalog/dino.rb (hunk 9 above)
Comment (zbright) [inline dino_catalog/dino.rb]: This will fail if you query on description since african dinos dont have that field

File: dino_catalog/dinodex.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  def big_dinos
+    @array_of_dinos.select {|dino| dino.heavier_than?(BIG)}
+  end
+
+  def small_dinos
+    @array_of_dinos.select {|dino| dino.lighter_than?(BIG)}
+  end
+
+  def carnivores
+    carnivores = @array_of_dinos.select { |dino|
```
Comment (djkotowski) [inline dino_catalog/dinodex.rb]: Unnecessary assignment.

File: dino_catalog/dinodex.rb [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+  def small_dinos
+    @array_of_dinos.select {|dino| dino.lighter_than?(BIG)}
+  end
+
+  def carnivores
+    carnivores = @array_of_dinos.select { |dino| 
+      dino.like?("DIET", "Carnivore") || 
+      dino.like?("DIET", "Piscivore") || 
+      dino.like?("DIET", "Insectivore")
```
Comment (djkotowski) [inline dino_catalog/dinodex.rb]: Could simplify to
```ruby
["Carnivore", "Piscivore", "Insectivore"].any? { |diet| dino.like?("DIET", diet) }
```

File: dino_catalog/Dinodex_Wizard.rb [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+  def self.prompt_user_multiple_constraints(constraint_array = [])
+    constraint_array << prompt_user_for_constraint
+    puts "add a constraint? (y/n)"
+    input = STDIN.gets.strip.downcase
+    if input == "y"
+      prompt_user_multiple_constraints(constraint_array)
+    else
+      constraint_array
+    end
```
Comment (djkotowski) [inline dino_catalog/Dinodex_Wizard.rb]: Use guard clauses to shrink this.

File: dino_catalog/dinoConverter.rb (hunk 6 above)
Comment (djkotowski) [inline dino_catalog/dinoConverter.rb]: Don't name variables after the data type.  Name them after what they represent.

File: dino_catalog/dinodex.rb (hunk 8 above)
Comment (djkotowski) [inline dino_catalog/dinodex.rb]:
```ruby
attr_reader :array_of_dinos
```

**Comment by djkotowski** [comment]
Chaining queries would be difficult here.  Instead of having Dinodex's filters return an array of filtered dinosaurs, why not have it return a new, filtered Dinodex?

**Comment by djkotowski** [comment]
Good job using `Enumerable` methods!  You'll find those will save you a lot of trouble.
Please look at the comments, refactor and resubmit.  Good work!

**Pull request closed without merging**